### Exploration of Hacker News Posts

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

In [1]:
#Import data and create a working list

from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [2]:
#Split the header from the data
headers = hn[0]
print(headers)
hn = hn[1:]
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [3]:
#Extracting the 'Ask HN' and 'Show'HN'data
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print('Number of ask posts: ', len(ask_posts))
print('Number of show posts: ', len(show_posts))
print('Number of other posts: ', len(other_posts))

Number of ask posts:  1744
Number of show posts:  1162
Number of other posts:  17194


In [4]:
#Find the count and average of each set

#Ask List
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])

avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average Ask Posts: ', round(avg_ask_comments,1))

#Show list
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])

avg_show_comments = total_show_comments / len(show_posts)
print('Average Show Posts: ', round(avg_show_comments, 1))

Average Ask Posts:  14.0
Average Show Posts:  10.3


The analysis is indicating that ask posts recieve more comments on average.

In [13]:
#Analysing the 'Ask Posts' and the time data
import datetime as dt

results_list = []

for row in ask_posts:
    results_list.append([row[6], int(row[4])])
    
comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in results_list:
    date = row[0]
    comment = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment

print('Counts: ', counts_by_hour)
print('Comments: ', comments_by_hour)


Counts:  {'03': 54, '22': 71, '06': 44, '23': 68, '12': 73, '17': 100, '09': 45, '21': 109, '20': 80, '02': 58, '14': 107, '00': 55, '15': 116, '19': 110, '11': 58, '13': 85, '04': 47, '08': 48, '05': 46, '18': 109, '10': 59, '07': 34, '01': 60, '16': 108}
Comments:  {'03': 421, '22': 479, '06': 397, '23': 543, '12': 687, '17': 1146, '09': 251, '21': 1745, '20': 1722, '02': 1381, '14': 1416, '00': 447, '15': 4477, '19': 1188, '11': 641, '13': 1253, '04': 337, '08': 492, '05': 464, '18': 1439, '10': 793, '07': 267, '01': 683, '16': 1814}


In [19]:
#Average number of posts per hour of the day

avg_hour = []

for hour in comments_by_hour:
    avg_hour.append([hour, comments_by_hour[hour] /
                     counts_by_hour[hour]])
    
avg_hour

[['03', 7.796296296296297],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['09', 5.5777777777777775],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['14', 13.233644859813085],
 ['00', 8.127272727272727],
 ['15', 38.5948275862069],
 ['19', 10.8],
 ['11', 11.051724137931034],
 ['13', 14.741176470588234],
 ['04', 7.170212765957447],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['18', 13.20183486238532],
 ['10', 13.440677966101696],
 ['07', 7.852941176470588],
 ['01', 11.383333333333333],
 ['16', 16.796296296296298]]

In [21]:
#Organise the data
swap_avg = []

for row in avg_hour:
    swap_avg.append([row[1], row[0]])
print(swap_avg)

sorted_swap = sorted(swap_avg, reverse = True)
print(sorted_swap)

[[7.796296296296297, '03'], [6.746478873239437, '22'], [9.022727272727273, '06'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [5.5777777777777775, '09'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.233644859813085, '14'], [8.127272727272727, '00'], [38.5948275862069, '15'], [10.8, '19'], [11.051724137931034, '11'], [14.741176470588234, '13'], [7.170212765957447, '04'], [10.25, '08'], [10.08695652173913, '05'], [13.20183486238532, '18'], [13.440677966101696, '10'], [7.852941176470588, '07'], [11.383333333333333, '01'], [16.796296296296298, '16']]
[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [

In [23]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

### Conclusion
In this project we determined which type of post and time receive the most comments on average. 

Based on the analysis it was an 'ask post' and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est). It should be noted that the data set we analyzed excluded posts without any comments. 

Therefore it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.